### PR TITLE
Unit test case for ocmp_powerSource

### PR DESCRIPTION
**Pull request by swateeshrivastava**
## Summary
Unit test case for ocmp_powerSource

Valid configuration has been taken from the /ec/platform/oc-sdr/cfg/OC_CONNECT_GBC.c for the test cases.
Common register file is created into /ec/test/fake/fake_SX1509_register.c).

## Test Plan
Run Unit Tests (make clean test) and verify that all tests pass including new test cases.
Run firmware build and verify no error is reported.

## Issues
#91 #207 #208

### CODE REVIEW

**Comment by swateeshrivastava** [comment]
Logs:
[ocmp_powerSource_logs.docx](https://github.com/Telecominfraproject/OpenCellular/files/2615300/ocmp_powerSource_logs.docx)

File: firmware/ec/test/include/test_powerSource.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#include "inc/devices/powerSource.h"
+#include "include/test_sx1509.h"
+#include <string.h>
+#include "unity.h"
+/* ======================== Constants & variables =========================== */
+#define I2C_BUS 5
+#define I2C_ADDR 0x71
+#define OC_EC_PWR_PRSNT_SOLAR_AUX 0x1E
+#define OC_EC_PWR_PRSNT_POE 0x55
+#define OC_EC_PWR_INVALID_PIN 0x57
```
Comment (jspectacular) [inline firmware/ec/test/include/test_powerSource.h]: Alphabetical order

File: firmware/ec/src/devices/ocmp_wrappers/ocmp_powerSource.c [hunk 2]
```diff
@@ -26,7 +26,11 @@ static bool _get_status(void *driver, unsigned int param_id, void *return_buf)
     return ret;
 }
 
-static ePostCode _probe(void *driver)
+/* Parameters are not used as this is just used to test assigning the
```
Comment (mdlewisfb) [inline firmware/ec/src/devices/ocmp_wrappers/ocmp_powerSource.c]: Can you update the comment to reflect the reason that there is an unused parameter in this function? I think it's just to match the defined function type, right?

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    SX1509_regs[SX1509_REG_DATA_A] = PWR_INT_EXT_BAT_DISABLE_SECOND_BYTE;
+    pwr_source_init();
+    pwr_get_source_info(&gbc_pwr_powerSource);
+    TEST_ASSERT_EQUAL(true, PWRSRC_fxnTable.cb_get_status(
+                                &gbc_pwr_powerSource, PWR_STAT_POE_AVAILABILITY,
+                                &powerStatus));
+    TEST_ASSERT_EQUAL(PWR_SRC_ACTIVE_AVAILABLE, powerStatus);
+
+    /* POE State Not Available */
+    powerStatus = PWR_SRC_NOT_AVAILABLE;
```
Comment (mdlewisfb) [inline firmware/ec/test/suites/Test_ocmp_powerSource.c]: This should be set to a value that doesn't match the expected final result tested on line 87.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    SX1509_regs[SX1509_REG_DATA_A] = PWR_INT_EXT_BAT_DISABLE_SECOND_BYTE;
+    pwr_source_init();
+    pwr_get_source_info(&gbc_pwr_powerSource);
+    TEST_ASSERT_EQUAL(true, PWRSRC_fxnTable.cb_get_status(
+                                &gbc_pwr_powerSource,
+                                PWR_STAT_POE_ACCESSIBILITY, &powerStatus));
+    TEST_ASSERT_EQUAL(PWR_SRC_ACTIVE_AVAILABLE, powerStatus);
+
+    /* POE State Not Accessible */
+    powerStatus = PWR_SRC_NOT_AVAILABLE;
```
Comment (mdlewisfb) [inline firmware/ec/test/suites/Test_ocmp_powerSource.c]: Same as above, should be set to not match the assert check on line 113.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+void test_init(void)
+{
+    PWR_GpioPins[OC_EC_PWR_PRSNT_SOLAR_AUX] = 0x1;
+    AlertData alert_data = {
+        .subsystem = 7,
+        .componentId = 1,
+        .deviceId = 0,
+    };
+    AlertData *alert_data_cp = malloc(sizeof(AlertData));
+    *alert_data_cp = alert_data;
```
Comment (jspectacular) [inline firmware/ec/test/suites/Test_ocmp_powerSource.c]: Do we want shallow copy?

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+void suite_tearDown(void)
+{
+    fake_I2C_deinit();
+}
+/* ================================ Tests =================================== */
+void test_init(void)
+{
+    PWR_GpioPins[OC_EC_PWR_PRSNT_SOLAR_AUX] = 0x1;
```
Comment (jspectacular) [inline firmware/ec/test/suites/Test_ocmp_powerSource.c]: Use #define for 0x01?

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+{
+    POSTData postData;
+    PWR_GpioConfig[OC_EC_PWR_PRSNT_SOLAR_AUX] = OCGPIO_CFG_OUTPUT;
+    PWR_GpioConfig[OC_EC_PWR_PRSNT_POE] = OCGPIO_CFG_OUTPUT;
+
+    TEST_ASSERT_EQUAL(POST_DEV_NOSTATUS, PWRSRC_fxnTable.cb_probe(
+                                             &gbc_pwr_powerSource, &postData));
+    TEST_ASSERT_EQUAL(OCGPIO_CFG_INPUT,
+                      PWR_GpioConfig[OC_EC_PWR_PRSNT_SOLAR_AUX]);
+    TEST_ASSERT_EQUAL(OCGPIO_CFG_INPUT, PWR_GpioConfig[OC_EC_PWR_PRSNT_POE]);
```
Comment (JoshuaJeyaraj) [inline firmware/ec/test/suites/Test_ocmp_powerSource.c]: can we test assertions for internal/external battery presence as well

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 8]
```diff
@@ -0,0 +1,292 @@
+#include "include/test_powerSource.h"
```
[the remark this answers is not in the document]
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: It is present into OpenCellular/firmware/ec/test/include/test_powerSource.h.

[the remark this answers is not in the document]
Comment (jspectacular) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Each file in repo needs to have its own header.

File: firmware/ec/test/include/test_powerSource.h (hunk 1 above)
Comment (swateeshrivastava) [reply firmware/ec/test/include/test_powerSource.h]: Code is updated.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c (hunk 3 above)
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Code is updated.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c (hunk 4 above)
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Code is updated.

File: firmware/ec/test/include/test_sx1509.h [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+#ifndef _TEST_SX1509_H
+#define _TEST_SX1509_H
+
+#include <stdint.h>
+
+/*Enumes are defined as per the SX1509 datasheet*/
```
[the remark this answers is not in the document]
Comment (swateeshrivastava) [reply firmware/ec/test/include/test_sx1509.h]: Code is updated.

File: firmware/ec/test/fake/fake_powerSource.c [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+#include "include/test_powerSource.h"
+#include <stdint.h>
+
+extern const OcGpio_FnTable GpioSX1509_fnTable;
+
+bool PWR_GpioPins[] = {
+    [OC_EC_PWR_PRSNT_SOLAR_AUX] = 0x1,
+    [OC_EC_PWR_PRSNT_POE] = 0x1,
```
[the remark this answers is not in the document]
Comment (swateeshrivastava) [reply firmware/ec/test/fake/fake_powerSource.c]: Code is updated.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Header is added.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c (hunk 5 above)
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: No, It is removed now.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c (hunk 6 above)
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Code is updated.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c (hunk 7 above)
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Code is updated.

File: firmware/ec/test/suites/Test_ocmp_powerSource.c [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    SX1509_regs[SX1509_REG_DATA_A] = PWR_INT_EXT_BAT_DISABLE_SECOND_BYTE;
+    pwr_source_init();
+    pwr_get_source_info(&gbc_pwr_powerSource);
+    TEST_ASSERT_EQUAL(true, PWRSRC_fxnTable.cb_get_status(
+                                &gbc_pwr_powerSource,
+                                PWR_STAT_SOLAR_AVAILABILITY, &powerStatus));
+    TEST_ASSERT_EQUAL(PWR_SRC_ACTIVE_AVAILABLE, powerStatus);
+
+    /* Solar State Not Available */
+    powerStatus = PWR_SRC_NOT_AVAILABLE;
```
[the remark this answers is not in the document]
Comment (swateeshrivastava) [reply firmware/ec/test/suites/Test_ocmp_powerSource.c]: Code is updated.

**Comment by swateeshrivastava** [comment]
Branch needs to be rebaselined.

I will take care of conflicts during merge.

File: firmware/ec/src/devices/powerSource.c [hunk 12]
```diff
@@ -275,56 +275,82 @@ pwr_process_get_status_parameters_data(ePower_StatusParamId paramIndex,
     switch (paramIndex) {
         case PWR_STAT_POE_AVAILABILITY: {
             if ((Power_SourceInfo[PWR_SRC_POE].state == PWR_SRC_ACTIVE) ||
-                (Power_SourceInfo[PWR_SRC_POE].state == PWR_SRC_AVAILABLE))
+                (Power_SourceInfo[PWR_SRC_POE].state == PWR_SRC_AVAILABLE)) {
```
Comment (JoshuaJeyaraj) [inline firmware/ec/src/devices/powerSource.c]: Lets init *pPowerStatusData = 0;  before switch case . . . . 
No need for the same code . . . .

**Comment by arulbose** [comment]
Looks good to me

**Pull request merged**